XWIKI-23080: Make LESS customizations impact CSS properties

Sereza7 opened the pull request:
# Jira URL

<!-- Add the link to the corresponding JIRA issue referenced in a commit message. Unless this is a [Misc] commit,
see https://dev.xwiki.org/xwiki/bin/view/Community/DevelopmentPractices#HRule:Don27tcreateunnecessaryissues
-->
https://jira.xwiki.org/browse/XWIKI-23080

# Changes

## Description

<!-- Describe the main changes brought in this PR. -->

* Added new lists of variables in the variablelist.vm template
* Removed them from the cssVariablesInit.css file
* Added a step at the end of the LESS variable initialisation to keep the CSS variables in line.

## Clarifications

<!-- Provide extra hints to make it easier to understand the PR. Those could be:
* Explanation of choices made in this PR
* Anchor towards extra resources needed to understand the context of this PR (e.g., a forum proposal).
* Links to other issues this issue depends on
-->

* As explained in the comments, I don't think we want to have perfect mapping for every variables. Some variables will be different in the CSS system and we can ensure that we don't break backwards compatibility if we only update them and not their LESS equivalents.
* As seen in the demo under here, the variables values set in SSXs do not impact the final value of the CSS variable. In the example shown at the end, the `@text-muted` LESS variable has two values defined at global level, but in different files. CSS properties can have only one global value (that would get overridden). So there's no way we can match 1-to-1 values perfectly here. IMO it's fine as things are now.

# Screenshots & Video

<!-- If this PR introduces any UI change, it's recommended to highlight it with before/after screenshots 
or even a screen recording for complex interactions. 
-->
Video demo of the feature. I took the `--text-muted` variable in this analysis. This is a variable that comes from bootstrap, we only provide one value for it in `cssVariablesInit.css`. I quickly added a special value (green) in the theme `lessCode`. We can see that it updated the end CSS value properly. When setting this variable value again in a SSX (red), we can see that the page still uses the `green` value in most places. The LESS value set in this SSX is used only on the compilation of this SSX. The SSX does not have any impact on the global scope CSS property.

FYI I'm using a CSS variable displayer macro available on a gist: https://gist.github.com/Sereza7/e4a78edc8c1c5865a44e84fc6a1b110d

https://github.com/user-attachments/assets/1d7534d5-529f-4cc4-8a2a-dd4e3f37a8b0



# Executed Tests

<!-- Especially important for regression fixes. 
Indicate how changes were tested (e.g., what maven commands were run to validate them).
-->

Manual tests, see above.
This is mostly front-end and barely covered, so for automated tests I just built the changes using: `mvn clean install -f xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources -Pquality`

# Expected merging strategy

* Prefers squash: Yes <!-- No — Explain why. -->
* Backport on branches:
  * None (same as https://github.com/xwiki/xwiki-platform/pull/3936)

michitux commented:
> When setting this variable value again in a SSX (red), we can see that the page still uses the green value in most places. The LESS value set in this SSX is used only on the compilation of this SSX. The SSX does not have any impact on the global scope CSS property.

Unless I missed something important, that's the same as in LESS, LESS values set in an SSX always only affect that SSX and nothing else.

What I'm wondering is: aren't most of the values set in `cssVariablesInit.css` obsolete now? Shouldn't we remove those values that are set directly to their LESS equivalents to reduce the size of the CSS as it causes network and parsing overhead? Also, it feels weird to have these default values defined in the CSS variable lists when they aren't used (from what I understand).

Sereza7 commented:
Adding on my previous message:

> I think it's important to make sure CSS properties can work (even if slightly reduced) with LESS completely down.

The LESS system is partially dependant on old colorthemes, having the CSS system dependant on the LESS system sounds to me like a really bad idea for technical debt.

michitux commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/less/variablesInit.vm:
```diff
@@ -192,3 +192,62 @@
+#foreach($variableName in $mappedVariables)
+  -/**/-$variableName: @$variableName;
+#end
+}
```
Isn't this the file that is included in all LESS templates? Wouldn't this mean that the CSS variables would be output as part of every SSX etc. that is processed with LESS? Or, asked differently where exactly is the output of these variables happening?

Sereza7 replied:
I'm not sure how the SSXs get their context for compilation, but the resulting CSS from this file is found only once on the page from the template use in `style.less.vm`

```
#template('less/variablesInit.vm')
```

Here is what the compilation result looks like with the test SSX I went with:
![image](https://github.com/user-attachments/assets/4e52b156-0357-4cb9-9cb2-a1a2225f7a58)

michitux replied:
The question is if it impacts the performance of LESS compilation. LESS compilation is already slow, this could have the potential to make it even slower. Could you try measuring the performance with and without your changes? Make sure that you don't measure the cache, force-reloading in the browser or appending URL parameters could help to avoid the cache.

Sereza7 replied:
I think that in order to force recompilation of the LESS, we also need to reset the instance or the colorTheme, it invalidates the server side cache.

I made sure to save a new version of the colortheme for the times shown under:

With he changes: 594ms rendering stylesheets.vm (read from the `debug` UI)
sandbox ssx: 2819ms
PDFExport SSX: 3053ms
Main ssx 3925ms

Without the changes:
682ms rendering stylesheets.vm
PDFExport SSX: 2689ms
sandbox ssx: 3666ms
Main ssx 3735ms

On this set of measures, the changes were slightly faster. It doesn't seem like the CSS variables in the context do much for performance. Note that those 70 CSS <-- LESS mappings might look like a lot, but there's a lot more LESS loaded and evaluated from all the skin files for each SSX :)


Sidenote: After just a hard refresh (without invalidating the server-side cache by saving the currently used flamingo colortheme), here is another quick set of number are:
Without the changes:
14ms rendering stylesheets.vm
sandbox ssx: 27ms
PDFExport SSX: 40ms
Main ssx 45ms

With the changes:
15ms rendering stylesheets.vm
sandbox ssx: 22ms
PDFExport SSX: 37ms
Main ssx 40ms

Both sets of measures are very similar still. AFAIK those numbers do not include any LESS compilation.

michitux commented:
Okay, even though it's not so nice (I would have preferred to have the CSS output in a file that is not included in all LESS compilations) the LESS compilation overhead is not a blocker for me. However, what I really don't like is that we now have a duplication of the initial values of the variables (e.g., from Bootstrap) that is not used at all. This seems like a sure way to have variables go out of sync and then when we remove the LESS fallback, we'll suddenly have different styles. Further, it creates a maintenance overhead (maintaining the values in two places) that has no real purpose except for having a fallback for when we remove LESS, which seems quite far away to me (several years at least I would say).

In my opinion (but I really don't want to decide this alone), we should remove at least the part of the CSS initialization that is unused and duplicates values. We can reconsider this when we work on having compatibility the other way around: Let color schemes indicate that they have a CSS variable section, not a LESS variable section, and generate LESS compatibility code based on the CSS variables. Then we could reverse this whole thing and only compile LESS compatibility variables if there is still some LESS code. However, at the moment, I think we're far from that step. To me, the first step is to migrate LESS code in XWiki to CSS only code such that the color scheme itself becomes the only LESS file that we still compile.

Sereza7 commented:
> In my opinion (but I really don't want to decide this alone), we should remove at least the part of the CSS initialization that is unused and duplicates values. 

I don't agree with this, IMO the pros outweight the cons:
* CONS: It makes it harder to maintain and fosters the creation of inconsistencies in the future. We hardly ever change values of existing variables. I did change a few for the Iceberg default to improve accessibility for https://jira.xwiki.org/browse/XWIKI-20680 , but AFAIK this was the last "big change" concerning variable default values. Most of the last changes were to introduce *new* values and not update *existing* values. New values can be added only in CSS now, there's no need to duplicate sources of truth.
* PROS: backup style if anything happens to what is done in this PR, and an easy way out once we can avoid LESS compilation altogether. I'll just add that right now, if the content of *one* LESS file fails compilation, *every* style from flamingo is down (since they are all put together for compilation, and our LESS compiler does not return any 'incomplete' stylesheet when it fails -- just the error logs). It makes debugging LESS a bit of a hassle. Having CSS rely solely on values set in LESS would mean that it'd behave the same. Native CSS is more resilient to bugs, we can completely have 'incomplete stylesheets'. They won't work perfectly of course, but unrelated styles will not be affected. From what I could see when trying to migrate, the worst thing that can happen when writing incorrect CSS is that anything declared after the mistake is removed from the styles (because whatever written closes the block early).

Here are a few screenshots to illustrate how unstable this looks when working with LESS:
Here's the home screen of a standard wiki using the default Iceberg theme.
![Screenshot from 2025-04-15 15-51-45](https://github.com/user-attachments/assets/98496e34-a807-4982-82df-5266ee67e390)
Then, as an admin, I added a single character to the end of the `@lessCode` field of Iceberg. It's just a star (*) and it's incorrect LESS to add it here. The compilation of all the LESS is now down, and my home screen looks like:
![Screenshot from 2025-04-15 15-52-03](https://github.com/user-attachments/assets/5c0015d6-e65c-4db7-a6de-9445bcb8b8a7)
This does not look like a refined way to fail at all. I think this would be quite intimidating for new devs trying to customize their instance to see that this one character (or the one mistake they just made when writing LESS) just broke everything. Note that on this instance, we have the cssVariablesInit.vm file running, so the CSS variables are still here, even if they are not used anywhere yet:
![Screenshot from 2025-04-15 15-52-24](https://github.com/user-attachments/assets/3cb23ca9-3a4e-458f-9d57-8da2b064f6a7)

Sereza7 commented:
Forgot to add a note here before:
We're discussing the direction we want to take on https://forum.xwiki.org/t/brainstorming-how-to-provide-values-to-css-properties/16851/3 . Setting this PR as a draft for now, it does not match what we're converging towards for a solution.

Sereza7 commented:
Summary of the changes:
* Removed the `html` font-size from bootstrap reset so that rem units can take into account the user preference.
* Moved the bootstrapVariable and XSVariable declaration from cssVariableInit.css to `variablelist.vm`
* Added the `login-form-max-width` to the list of new variables, because it was forgotten.
* Renamed miscellaneous `variablesInit.vm` velocimacros and their uses so that they follow our naming convention for non API velocimacros (added an underscore at the start)
* Factorized the initialization of xsFromOldColorThemeVariables into `variablelist.vm`
* Added LESS -> CSS variable values, taking priority over the defaults

Sereza7 commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/variablelist.vm:
```diff
@@ -95,13 +95,123 @@
+## XS specific variables
+############################################
+## List of all the variables set in XWiki standard for various purposes.
+#set($xsVariables = {
```
Addressed in https://github.com/xwiki/xwiki-platform/pull/4037/commits/d3fbc4686d8f02dc5c29cfbf433bef8a8be995fd :+1:
